url2df0f7c811_przemekgalazka_legacycode_tgz-pJ8-pl_cc_core_cmd_CommandJ8/src/ccproxy/AgentConnectionList.java

```diff
@@ -0,0 +1,108 @@
+package ccproxy;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import pl.cc.SystemCoreState;
+import pl.cc.core.cmd.Command;
+
+public class AgentConnectionList {
+	private List<AgentConnection> aclist = Collections.synchronizedList(new ArrayList<AgentConnection>());
+	SystemCoreState coreState;
+	
+	public AgentConnectionList(SystemCoreState coreState) {
+		this.coreState = coreState;
+	}
+
+	public synchronized int size() {
+		return aclist.size();
+	}
+
+	public synchronized AgentConnection get(int i) {
+		return aclist.get(i);
+	}
+
+	public synchronized boolean contains(Object o) {
+		return aclist.contains(o);
+	}
+
+	public synchronized void add(AgentConnection ac) {
+		aclist.add(ac);
+	}
+
+	public synchronized void remove(AgentConnection ac) {
+		aclist.remove(ac);
+	}
+
+	public synchronized void remove(int i) {
+		aclist.remove(i);
+	}
+
+	public synchronized AgentConnection getByIP(String ip) {
+		for (int i = 0; i < aclist.size(); i++) {
+			if (aclist.get(i).ip.equals(ip))
+				return aclist.get(i);
+		}
+		return null;
+	}
+
+	public synchronized AgentConnection getAgentConnectionByNumer(String number) {
+		for (int i = 0; i < aclist.size(); i++) {
+			if (aclist.get(i).agentNumber.equals(number))
+				// //System.out.println("dupadapduapdsa");
+				return aclist.get(i);
+		}
+		return null;
+	}
+
+	public synchronized AgentConnection getAgentConnectionByExten(String exten) {
+		for (int i = 0; i < aclist.size(); i++) {
+			if (aclist.get(i).agentExten.equals(exten))
+				// //System.out.println("dupadapduapdsa");
+				return aclist.get(i);
+		}
+		return null;
+	}
+
+    public synchronized void pushSupervisorMessage(String message) {
+        for (int i = 0; i < aclist.size(); i++) {
+            if (aclist.get(i).supervisor){
+                aclist.get(i).out.println(message);
+            }
+        }
+    }
+
+    public synchronized void pushMessage(String message) {
+		for (int i = 0; i < aclist.size(); i++) {
+			aclist.get(i).out.println(message);
+		}
+	}
+
+	public void pushMessage(String message, AgentConnection ac) {
+		ac.out.println(message);
+	}
+
+	public void pushQueueMessage(String message) {
+		for (int i = 0; i < aclist.size(); i++) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **87**. At this line the following field is not protected by synchronization: `aclist`. Your code typically protects this field with synchronization on object: `AgentConnectionList.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `AgentConnectionList.this`. For example, like at lines: **19**, **35**, **23**, **39**, **27**, **31**, **78**, **77**, **69**, **70**, and **71**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **19**, **23**, **27**, **31**, **35**, **39**, **43**, **44**, **45**, **51**, **52**, **54**, **60**, **61**, **63**, **69**, **70**, **71**, **77**, **78**, **102**, and **103**. This is a call to `size()`. You are already protecting `size()`. You are protecting these operations at lines: **19**.